src/ExpressionEvaluator/Core/Source/ResultProvider/Expansion/TupleExpansion.cs

```diff
@@ -0,0 +1,266 @@
+
+            internal Field(
+                TypeAndCustomInfo declaringTypeAndInfo,
+                TypeAndCustomInfo fieldTypeAndInfo,
+                FieldInfo fieldInfo,
+                string name,
+                Field parent)
+            {
+                Debug.Assert(declaringTypeAndInfo.ClrType != null);
+                Debug.Assert(fieldTypeAndInfo.ClrType != null);
+                Debug.Assert(fieldInfo != null);
+                Debug.Assert(name != null);
+                Debug.Assert(declaringTypeAndInfo.Type.Equals(fieldInfo.DeclaringType));
+                Debug.Assert(fieldTypeAndInfo.Type.Equals(fieldInfo.FieldType));
+                Debug.Assert(parent == null || parent.FieldInfo.FieldType.Equals(fieldInfo.DeclaringType));
+
+                DeclaringTypeAndInfo = declaringTypeAndInfo;
+                FieldTypeAndInfo = fieldTypeAndInfo;
+                FieldInfo = fieldInfo;
+                Name = name;
+                Parent = parent;
+            }
+        }
+
+        private ReadOnlyCollection<Field> GetFields()
+        {
+            if (_lazyFields == null)
+            {
+                _lazyFields = GetFields(_typeAndInfo, _cardinality);
+            }
+            return _lazyFields;
+        }
+
+        private static ReadOnlyCollection<Field> GetFields(TypeAndCustomInfo declaringTypeAndInfo, int cardinality)
+        {
+            Debug.Assert(declaringTypeAndInfo.Type.GetTupleCardinalityIfAny() == cardinality);
+
+            var appDomain = declaringTypeAndInfo.ClrType.AppDomain;
+
+            var customTypeInfoMap = CustomTypeInfoTypeArgumentMap.Create(declaringTypeAndInfo);
+            var tupleElementNames = customTypeInfoMap.TupleElementNames;
+
+            var builder = ArrayBuilder<Field>.GetInstance();
+            Field parent = null;
+            int offset = 0;
+
+            while (true)
+            {
+                var declaringType = declaringTypeAndInfo.Type;
+                int n = Math.Min(cardinality, TypeHelpers.TupleFieldRestPosition - 1);
+                for (int index = 0; index < n; index++)
+                {
+                    var field = declaringType.GetTupleField(TypeHelpers.GetTupleFieldName(index));
+                    if (field == null)
+                    {
+                        // Ignore missing fields.
+                        continue;
+                    }
+
+                    var fieldTypeAndInfo = GetTupleFieldTypeAndInfo(appDomain, field, customTypeInfoMap);
+                    var name = CustomTypeInfo.GetTupleElementNameIfAny(tupleElementNames, offset + index);
+                    if (name != null)
+                    {
+                        builder.Add(new Field(declaringTypeAndInfo, fieldTypeAndInfo, field, name, parent));
+                    }
+                    builder.Add(new Field(declaringTypeAndInfo, fieldTypeAndInfo, field, TypeHelpers.GetTupleFieldName(offset + index), parent));
```

Comment:
>TypeHelpers.GetTupleFieldName(offset + index) [](start = 89, length = 45)

Looks like this is used twice (see above).  Worth caching into a local to avoid the allocations?